ipaserver/install/dsinstance.py

```diff
@@ -528,68 +526,83 @@ def __setup_sub_dict(self):
+            ESCAPED_SUFFIX=str(self.suffix),
+            GROUP=DS_GROUP,
+            IDRANGE_SIZE=idrange_size,
+            DOMAIN_LEVEL=self.domainlevel,
+            MAX_DOMAIN_LEVEL=constants.MAX_DOMAIN_LEVEL,
+            MIN_DOMAIN_LEVEL=constants.MIN_DOMAIN_LEVEL,
+            STRIP_ATTRS=" ".join(replication.STRIP_ATTRS),
+            EXCLUDES='(objectclass=*) $ EXCLUDE ' +
+            ' '.join(replication.EXCLUDES),
+            TOTAL_EXCLUDES='(objectclass=*) $ EXCLUDE ' +
+            ' '.join(replication.TOTAL_EXCLUDES),
+        )
 
     def __create_instance(self):
-        pent = pwd.getpwnam(DS_USER)
-
         self.backup_state("serverid", self.serverid)
         self.fstore.backup_file(paths.SYSCONFIG_DIRSRV)
 
-        self.sub_dict['BASEDC'] = self.realm.split('.')[0].lower()
-        base_txt = ipautil.template_str(BASE_TEMPLATE, self.sub_dict)
-        logger.debug("%s", base_txt)
-
-        target_fname = paths.DIRSRV_BOOT_LDIF
-        base_fd = open(target_fname, "w")
-        base_fd.write(base_txt)
-        base_fd.close()
-
-        # Must be readable for dirsrv
-        os.chmod(target_fname, 0o440)
-        os.chown(target_fname, pent.pw_uid, pent.pw_gid)
-
-        inf_txt = ipautil.template_str(INF_TEMPLATE, self.sub_dict)
-        logger.debug("writing inf template")
-        inf_fd = ipautil.write_tmp_file(inf_txt)
-        inf_txt = re.sub(r"RootDNPwd=.*\n", "", inf_txt)
-        logger.debug("%s", inf_txt)
-        args = [
-            paths.SETUP_DS_PL, "--silent",
-            "--logfile", "-",
-            "-f", inf_fd.name,
-        ]
-        logger.debug("calling setup-ds.pl")
-        try:
-            ipautil.run(args)
-            logger.debug("completed creating DS instance")
-        except ipautil.CalledProcessError as e:
-            raise RuntimeError("failed to create DS instance %s" % e)
+        # The new installer is api driven. We can pass it a log function
+        # and it will use it. Because of this, we can pass verbose true,
+        # and allow our logger to control the display based on level.
+        sds = SetupDs(verbose=True, dryrun=False, log=logger)
+
+        # General environmental options.
+        general_options = General2Base(logger)
+        general_options.set('strict_host_checking', False)
+        # Check that our requested configuration is actually valid ...
+        general_options.verify()
+        general = general_options.collect()
+
+        # Slapd options, ie instance name.
+        slapd_options = Slapd2Base(logger)
+        slapd_options.set('instance_name', self.serverid)
+        slapd_options.set('root_password', self.dm_password)
+        slapd_options.verify()
+        slapd = slapd_options.collect()
+
+        # Create userroot. Note that the new install does NOT
+        # create sample entries, so this is *empty*.
+        userroot = {
+            'cn': 'userRoot',
+            'nsslapd-suffix': self.suffix.ldap_text()
+        }
+
```

Comment:
Adding backends now works in the new installer, and you can add sample entries if you want to, but the backend object should look like:

userroot = {
    'name': 'userroot', 
    'suffix': self.suffix.ldap_text(),
    'sample_entries': 'no'}

Reply:
Ok, addressed.